tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc

```diff
@@ -444,27 +482,169 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
                               const HloModuleConfig& hlo_module_config,
                               const string& libdevice_dir_path) {
   static std::once_flag backend_init_flag;
-  std::call_once(backend_init_flag, GPUBackendInit, hlo_module_config);
+  std::call_once(backend_init_flag, NVPTXBackendInit, hlo_module_config);
 
   string ptx;
+  std::unique_ptr<llvm::TargetMachine> target_machine;
   {
     tensorflow::profiler::TraceMe activity(
         [&] { return absl::StrCat("Compiling IR:", module->getName().str()); },
         tensorflow::profiler::TraceMeLevel::kInfo);
     XLA_SCOPED_LOGGING_TIMER("Compile module " + module->getName().str());
+
+    // If the module has no functions or globals, there's nothing to compile.
+    // Just return an empty string.
+    if (module->empty() && module->global_empty()) {
+      VLOG(2) << "Module '" << module->getName().str()
+              << "' is empty. Skipping compilation.";
+      return string();
+    }
+
     TF_ASSIGN_OR_RETURN(
-        ptx, CompileModuleToPtx(module, compute_capability, hlo_module_config,
-                                libdevice_dir_path));
+        target_machine,
+        LinkAndOptimizeModule(module, compute_capability, hlo_module_config,
+                              libdevice_dir_path));
+    TF_ASSIGN_OR_RETURN(ptx, EmitModuleToPTX(module, target_machine.get()));
   }
   return ptx;
 }
 
+}  // namespace nvptx
+
+namespace amdgpu {
+
+// Gets the ROCm-Device-Libs filenames for a particular AMDGPU version.
+static std::vector<string> GetROCDLPaths(int amdgpu_version,
+                                         const string& rocdl_dir_path) {
+  // AMDGPU version-neutral bitcodes.
+  std::vector<string> rocdl_filename_vector{
+      "hc.amdgcn.bc",
+      "opencl.amdgcn.bc",
+      "ocml.amdgcn.bc",
+      "ockl.amdgcn.bc",
+      "oclc_finite_only_off.amdgcn.bc",
+      "oclc_daz_opt_off.amdgcn.bc",
+      "oclc_correctly_rounded_sqrt_on.amdgcn.bc",
+      "oclc_unsafe_math_off.amdgcn.bc"};
+
+  // Construct full path to ROCDL bitcode libraries.
+  std::vector<string> result;
+  for (auto& filename : rocdl_filename_vector) {
+    result.push_back(tensorflow::io::JoinPath(rocdl_dir_path, filename));
+  }
+
+  // Add AMDGPU version-specific bitcodes.
+  result.push_back(tensorflow::io::JoinPath(
+      rocdl_dir_path, tensorflow::strings::StrCat(
+                          "oclc_isa_version_", amdgpu_version, ".amdgcn.bc")));
+  return std::move(result);
+}
+
+// Links ROCm-Device-Libs into the given module if the module needs it.
+Status LinkROCDLIfNecessary(llvm::Module* module, int amdgpu_version,
+                            const string& rocdl_dir_path) {
+  if (!CouldNeedDeviceBitcode(*module)) {
+    return tensorflow::Status::OK();
+  }
+
+  return LinkWithBitcodeVector(module,
+                               GetROCDLPaths(amdgpu_version, rocdl_dir_path));
+}
+
+StatusOr<std::unique_ptr<llvm::TargetMachine>> LinkAndOptimizeModule(
```

Comment:
In the previous round I asked you to remove the repetitive
    if (isNVPTX) { ... } else { ... }
pattern, because it makes the code hard to read. You are right when you say that this introduced code duplication.

However, I think we can get the best of both worlds as follows: Let's extract target-specific parts of this function, i.e. linking device libs and generating the target machine. What remains will be a target-independent LinkAndOptimizeModule function that can be called with the right inputs form CompileToPtx and the AMDGPU equivalent. Does that make sense?

Reply:
@thomasjoerg Ok, let me take a stab try extracting common logic from the 2 and pass in target-specific logic as callable functions.

On NVPTX side, target-specific logic are:
- link with libdevice if necessary
- add module flag
- add function attribute
- set target triple
- get target machine
- add optimization passes and set inliner threshold

On AMDGPU side, target-specific logic are:
- link with ROCm-Device-Libs if necessary
- set target triple
- get target machine
- add optimization passes and set inliner threshold

Reply:
@thomasjoerg I've submitted a new version. Please kindly review it again. Thanks!